### PR TITLE
Bluetooth: shell: allow multiple parameters for bt hci-cmd

### PR DESCRIPTION
**Pull request by kruithofa**
The hci-cmd currently accepts only 1 byte parameter. This
disables use of commands that requires more than 1 byte,
for ex. the feature exchange that requires a 16 bit parameter.
This PR allows up to 32 bytes, which should be enough for
all HCI commands ref. BT V5.2 spec chapter 7.8

Signed-off-by: Andries Kruithof <Andries.Kruithof@nordicsemi.no>

### CODE REVIEW

**Review by joerchan — APPROVED** [review]

**Review by jhedberg — CHANGES_REQUESTED** [review]
HCI_LE_Generate_DHKey can take up to 65 bytes of parameters (Part E section 7.8.37 LE Generate DHKey command), so that might be a more reasonable maximum

**Comment by joerchan** [comment]
> HCI_LE_Generate_DHKey can take up to 65 bytes of parameters (Part E section 7.8.37 LE Generate DHKey command), so that might be a more reasonable maximum

Maybe not have the bytes as space-separated and instead use `bin2hex`, that way we don't have to have a max length.
`cmd_adv_data` does it like this.

**Comment by kruithofa** [comment]
> > HCI_LE_Generate_DHKey can take up to 65 bytes of parameters (Part E section 7.8.37 LE Generate DHKey command), so that might be a more reasonable maximum
> 
> Maybe not have the bytes as space-separated and instead use `bin2hex`, that way we don't have to have a max length.
> `cmd_adv_data` does it like this.

Ok, I suppose using bin2hex is more consistent with the other commands.
Note that we still have a max. length of 65 bytes since we need to reserve bufferspace; we could do dynamic allocation but that makes the code unnecessarily complicated.

File: subsys/bluetooth/shell/bt.c [hunk 1]
```diff
@@ -453,19 +455,34 @@ static int cmd_hci_cmd(const struct shell *shell, size_t argc, char *argv[])
...
+	hex_data_len = 0;
 	ogf = strtoul(argv[1], NULL, 16);
 	ocf = strtoul(argv[2], NULL, 16);
 
 	if (argc > 3) {
+		size_t len;
 		size_t i;
 
-		buf = bt_hci_cmd_create(BT_OP(ogf, ocf), argc - 3);
+		if (strlen(argv[3]) > 2*HCI_CMD_MAX_PARAM) {
```
Comment (jhedberg) [inline subsys/bluetooth/shell/bt.c:471]: Minor: `2 * HCI_CMD_MAX_PARAM`, i.e. there should be a space before and after the `*` operator

File: subsys/bluetooth/shell/bt.c [hunk 2]
```diff
@@ -56,6 +56,8 @@ static struct bt_le_oob oob_remote;
 
 #define KEY_STR_LEN 33
 
+#define HCI_CMD_MAX_PARAM 65
```
Comment (jhedberg) [inline subsys/bluetooth/shell/bt.c:59]: Maybe a code comment above this to explain how the value is chosen? (e.g. a mention of the DHKey command)

**Review by jhedberg — CHANGES_REQUESTED** [review]
Two tiny things - after that I'm happy to approve

File: subsys/bluetooth/shell/bt.c [hunk 3]
```diff
@@ -453,19 +455,34 @@ static int cmd_hci_cmd(const struct shell *shell, size_t argc, char *argv[])
...
+		if (!len) {
+			shell_error(shell, "HCI command illegal data field\n");
+			return -ENOEXEC;
+		}
 
-		for (i = 3; i < argc; i++) {
-			net_buf_add_u8(buf, strtoul(argv[i], NULL, 16));
+		buf = bt_hci_cmd_create(BT_OP(ogf, ocf), len);
+		for (i = 0; i < len; i++) {
+			net_buf_add_u8(buf, hex_data[i]);
```
Comment (joerchan) [inline subsys/bluetooth/shell/bt.c:485]: I think net_buf_add_mem could be used here.

**Comment by carlescufi** [comment]
@kruithofa will you follow-up on this one?

File: subsys/bluetooth/shell/bt.c (hunk 1 above)
Comment (kruithofa) [reply subsys/bluetooth/shell/bt.c]: Done

File: subsys/bluetooth/shell/bt.c (hunk 2 above)
Comment (kruithofa) [reply subsys/bluetooth/shell/bt.c:63]: Comment added

File: subsys/bluetooth/shell/bt.c (hunk 3 above)
Comment (kruithofa) [reply subsys/bluetooth/shell/bt.c]: Replaced with net_buf_add_mem

**Comment by kruithofa** [comment]
Rebased after the change to C99 code-style

**Review by joerchan — CHANGES_REQUESTED** [review]

**Comment by kruithofa** [comment]
DNM until shippable is OK

File: subsys/bluetooth/shell/bt.c [hunk 4]
```diff
@@ -455,18 +461,30 @@ static int cmd_hci_cmd(const struct shell *shell, size_t argc, char *argv[])
...
 
-		buf = bt_hci_cmd_create(BT_OP(ogf, ocf), argc - 3);
+		if (strlen(argv[3]) > 2 * HCI_CMD_MAX_PARAM) {
+			shell_error(shell, "Data field too large\n");
+			return -ENOEXEC;
+		}
 
-		for (i = 3; i < argc; i++) {
-			net_buf_add_u8(buf, strtoul(argv[i], NULL, 16));
+		len = hex2bin(argv[3], strlen(argv[3]), &hex_data[hex_data_len],
```
Comment (joerchan) [inline subsys/bluetooth/shell/bt.c:479]: hex_data_len is just zero, you could get rid of this variable IMO.

**Review by joerchan — APPROVED** [review]

File: subsys/bluetooth/shell/bt.c (hunk 4 above)
Comment (kruithofa) [reply subsys/bluetooth/shell/bt.c:479]: Yes, considered that ,but left it, in case we want to split up the parameters send

**Comment by kruithofa** [comment]
The failing test (kernel/lifo/lifo_usage) seem to fail intermittently both in this PR and in the master, therefor removed DNM label

**Comment by joerchan** [comment]
Open and close to rerun failed shippable.

**Review by jhedberg — APPROVED** [review]

**Pull request merged**